app/controllers/posts_controller.rb

```diff
@@ -26,12 +27,26 @@ def edit; end
 
   def update
     if @post.update(post_params)
-      redirect_to posts_url, notice: 'Post updated successfully!'
+      flash[:success] = 'Post updated successfully!'
+      redirect_to posts_url
     else
       render :edit
     end
   end
 
+  def destroy
+    @post.discard
+    flash[:alert] = "Post Removed. #{view_context.link_to('Undo', restore_post_path(@post))}".html_safe
+    redirect_to posts_url
+  end
+
+  def restore
+    @post.undiscard
+    flash[:success] = 'Post recovered successfully!'
+    redirect_to posts_url
```

Comment:
If you've just restored a post, you probably want to be redirected to that post directly.